README.md
ver. 0.2.00 

[![Build Status](https://travis-ci.org/tkych/cl-spark.png)](https://travis-ci.org/tkych/cl-spark)
[cl-test-grid](http://common-lisp.net/project/cl-test-grid/library/cl-spark.html)


cl-spark
========

cl-spark generates a [sparkline][Sparkline theory and practice by Edward Tufte] string for a list of numbers.
It is a common lisp implementation of Zach Holman's [spark][spark] and Gil Gonçalves' [vspark][vspark] with little extension.

The goal of cl-spark is quick data visualization for:

 1. quick-and-dirty tasks
    (e.g. check the number of http-requests, [github-commits][github-commits], temperature, etc.),
 2. deciding whether the data is worth to real data mining
    (e.g. [monitoring Fukushima Nuclear Power Station][fukushima-monitor]),
 3. grasping the secret aesthetic beauty
    (e.g. visualizing the fibonacci sequence, you can feel the golden ratio).


Note:

 * The character encoding of cl-spark must be `utf-8`.
   If your lisp is *clozure*, you need [command-line option][ccl-option] `$ ccl -K utf-8`
   when you start lisp on the shell.


[spark]:  https://github.com/holman/spark
[vspark]: https://github.com/LuRsT/vspark
[Sparkline theory and practice by Edward Tufte]: http://www.edwardtufte.com/bboard/q-and-a-fetch-msg?msg_id=0001OR
[github-commits]: https://gist.github.com/tkych/6593401
[fukushima-monitor]: https://gist.github.com/tkych/6509285
[ccl-option]: http://ccl.clozure.com/manual/chapter2.5.html#Command-Line-Options


Depends-on
----------

 - NIL


Installation
------------

#### cl-test-grid results:

 - http://common-lisp.net/project/cl-test-grid/library/cl-spark.html


#### Auto:

 0. CL-REPL> `(ql:quickload :cl-spark)`


#### Manual:

 0. SHELL$   `git clone https://github.com/tkych/cl-spark`
 1. CL-REPL> `(push #p"/path-to-cl-spark/cl-spark/" asdf:*central-registry*)`
 2. CL-REPL> `(ql:quickload :cl-spark)` or `(asdf:load-system :cl-spark)`


Examples
--------

    ;;; Spark

    (spark '(1 1 2 3 5 8)) => "▁▁▂▃▅█"

    ;; float, minus
    (spark '(1 0 1 0))    => "█▁█▁"
    (spark '(1 0 1 0 .5)) => "█▁█▁▄"
    (spark '(1 0 1 0 -1)) => "█▄█▄▁"

    ;; min, max
    (spark '(0 30 55 80 33 150))                 => "▁▂▃▅▂█"
    (spark '(0 30 55 80 33 150) :min -100)       => "▃▄▅▆▄█"
    (spark '(0 30 55 80 33 150) :max 50)         => "▁▅██▅█"
    (spark '(0 30 55 80 33 150) :min 30 :max 80) => "▁▁▄█▁█"

    ;; key
    (spark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4))))
    => "▄▆█▆▄▂▁▂▄"
    (spark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (cos (* x pi 1/4))))
    => "█▆▄▂▁▂▄▆█"

    ;; in function
    (defun look-bits (n)
      (spark (map 'list #'digit-char-p (write-to-string n :base 2))))

    (look-bits 42) => "█▁█▁█▁"
    (look-bits 43) => "█▁█▁██"
    (look-bits 44) => "█▁██▁▁"
    (look-bits 45) => "█▁██▁█"

    ;; *ticks*
    (defvar ternary '(-1 0 1 -1 1 0 -1 1 -1))

    (spark ternary)              => "▁▄█▁█▄▁█▁"

    (let ((*ticks* #(#\_ #\- #\¯)))
      (spark ternary))           => "_-¯_¯-_¯_"

    (let ((*ticks* #(#\▄ #\⎯ #\▀)))
      (spark ternary))           => "▄⎯▀▄▀⎯▄▀▄"


    ;;; Vspark

    ;; Life expectancy by WHO region, 2011, bothsexes
    ;; see. http://apps.who.int/gho/data/view.main.690
    (defvar life-expectancies '(("Africa" 56)
                                ("Americans" 76)
                                ("South-East Asia" 67)
                                ("Europe" 76)
                                ("Eastern Mediterranean" 68)
                                ("Western Pacific" 76)
                                ("Global" 70)))

    (vspark life-expectancies :key #'second)
    =>
    "
    56                      66                      76
    ˫-----------------------+------------------------˧
    ▏
    ██████████████████████████████████████████████████
    ███████████████████████████▌
    ██████████████████████████████████████████████████
    ██████████████████████████████▏
    ██████████████████████████████████████████████████
    ███████████████████████████████████▏
    "

    (vspark life-expectancies :key #'second
                              :min 50 :max 80
                              :labels (mapcar #'first life-expectancies)
                              :title "Life Expectancy")
    =>
    "
                     Life Expectancy                  
                          50           65           80
                          ˫------------+-------------˧
                   Africa █████▋
                Americans ████████████████████████▎
          South-East Asia ███████████████▉
                   Europe ████████████████████████▎
    Eastern Mediterranean ████████████████▊
          Western Pacific ████████████████████████▎
                   Global ██████████████████▋
    "

    ;; labels, size
    (vspark '(1 0 .5) :labels '("on" "off" "unknown") :size 1)
    =>
    "
         on █
        off ▏
    unknown ▌
    "

    (vspark '(1 0 .5) :labels '("on" "off") :size 1)
    =>
    "
     on █
    off ▏
        ▌
    "

    (vspark '(1 0) :labels '("on" "off" "unknown") :size 1)
    =>
    "
     on █
    off ▏
    "

    ;; auto-scale
    (vspark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4)))
                                 :size 20)
    =>
    "
    -1.0     0.0     1.0
    ˫--------+---------˧
    ██████████▏
    █████████████████▏
    ████████████████████
    █████████████████▏
    ██████████▏
    ██▉
    ▏
    ██▉
    █████████▉
    "

    (vspark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4)))
                                 :size 10)
    =>
    "
    -1.0   1.0
    ˫--------˧
    █████▏
    ████████▏
    ██████████
    ████████▏
    █████▏
    █▏
    ▏
    █▏
    ████▏
    "
    (vspark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4)))
                                 :size 7)
    =>
    "
    ███▌
    █████▉
    ███████
    █████▉
    ███▌
    █▏
    ▏
    █▏
    ███▌
    "


For more examples, see cl-spark/test.lisp


Referece Manual
---------------

#### [Function] SPARK _numbers_ _&key_ _min_ _max_ _key_

Generates a sparkline string for a list of real numbers.

  * _numbers_ is a data, list of real-numbers.
  * _min_ is lower bound of output, either `nil` or  real-number (default is `NIL`, the minimum value of the data).
  * _max_ is upper bound of output, either `nil` or  real-number (default is `NIL`, the maximum value of the data).
  * _key_ is a function for preparing data.


#### [Special Variable] \*TICKS\*

A simple-vector of characters for representation of sparklines.
Default is `#(#\▁ #\▂ #\▃ #\▄ #\▅ #\▆ #\▇ #\█)`.


#### [Function] VSPARK _numbers_ _&key_ _min_ _max_ _key_ (_size_ 50) _title_ _labels_ (_scale?_ t) (_newline?_ t)

Generates a vartical sparkline string for a list of real numbers.

  * _numbers_ is a data, list of real-numbers.
  * _min_ is lower bound of output, either `nil` or real-number (default is `NIL`, the minimum value of the data).
  * _max_ is upper bound of output, either `nil` or real-number (default is `NIL`, the maximum value of the data).
  * _key_ is a function for preparing data.
  * _size_ is a maximum number of output columns (contains label), integer (default is 50).
  * _title_ is a title for data, string or nil. If title is too big for size, then not print.
  * _labels_ is a labels for data, list.
  * _scale?_ is a boolean (default is `T`). If T, output graph with scale for easy to see.
    If string length of min and max is too big for size, then not print scale.
  * _newline?_ is a boolean (default is `T`). If T, output graph with newlines for easy to see.


#### [Special Variable] \*VTICKS\*

A simple-vector of characters for representation of vartical sparklines.
Default is `#(#\▏ #\▎ #\▍ #\▌ #\▋ #\▊ #\▉ #\█)`.


Author, License, Copyright
--------------------------

- Takaya OCHIAI  <#.(reverse "moc.liamg@lper.hcykt")>

- MIT License

- Copyright (C) 2013-2014 Takaya OCHIAI


CHANGELOG
cl-spark/CHANGELOG

Version 0.2.00
2014-08-12
REFACTOR: VSPARK, GENERATE-SCALE, to be portable.
ADD: travis.yaml for Travis CI.

Version 0.1.13
2014-06-15
ADD: cl-spark-test.asd, spark-test.lisp
REFACTOR: VSPARK
FIX: key's type (or null function) -> (or symbol function)

Version 0.1.02
2013-09-16
RENAME: inf -> min, sup -> max.

Version 0.1.01
2013-09-13
FIX: VSPARK: If all values are under the inf,
             then sparkline becomes incorrect.
e.g.
(vspark '(1 2) :inf 10)
=>
"
9                      10                       11
˫-----------------------+------------------------˧
█████████████████████████▏
█████████████████████████▏
"

Fixed:
=>
"
10                     10.5                     11
˫-----------------------+------------------------˧
▏
▏
"

Version 0.1.00
2013-09-10
Initial Release

Version 0.0.00
2013-09-06
Project Start


run-tests.lisp
#!/usr/bin/sbcl --script

;;;; cl-spark/run-tests.lisp

;; Copyright (c) 2013 Takaya OCHIAI <tkych.repl@gmail.com>
;; This software is released under the MIT License.
;; For more details, see cl-spark/LICENSE

;;====================================================================
;; Test for CL-SPARK with Travis-CI
;;====================================================================

(load "quicklisp.lisp")

(quicklisp-quickstart:install)

(ql:quickload :cl-spark-test)

(if (cl-spark-test:run-tests)
    (sb-ext:quit :unix-status 0)
    (sb-ext:quit :unix-status 1))


;;====================================================================


spark.lisp
;;;; cl-spark/spark.lisp

;; Copyright (c) 2013-2014 Takaya OCHIAI <tkych.repl@gmail.com>
;; This software is released under the MIT License.
;; For more details, see cl-spark/LICENSE

;;====================================================================
;; CL-SPARK
;;====================================================================

(in-package :cl-user)
(defpackage #:cl-spark
  (:nicknames #:spark)
  (:use :cl)
  (:export #:spark  #:*ticks*
           #:vspark #:*vticks*))
(in-package #:cl-spark)


;;--------------------------------------------------------------------
;; Utils
;;--------------------------------------------------------------------

(defun string-concat (&rest strings)
  (with-output-to-string (s)
    (dolist (string strings)
      (princ string s))))

(defmacro aif (test true-clause &optional false-clause)
  `(let ((it ,test))
     (if it
         ,true-clause
         ,false-clause)))

(defmacro awhen (test &body body)
  `(let ((it ,test))
     (when it
       ,@body)))


;;--------------------------------------------------------------------
;; Type
;;--------------------------------------------------------------------

(eval-when (:compile-toplevel :load-toplevel :execute)
  (defun at-least-two-chars-p (x)
    (and (simple-vector-p x)
         (<= 2 (length x))
         (every #'characterp x))))

(deftype <ticks> ()
  `(and simple-vector (satisfies at-least-two-chars-p)))

(declaim (type <ticks> *ticks* *vticks*))


;;--------------------------------------------------------------------
;; Spark
;;--------------------------------------------------------------------
;; (vector #\▁ #\▂ #\▃ #\▄ #\▅ #\▆ #\▇ #\█)

;; (code-char 9600) => #\UPPER_HALF_BLOCK           <=> #\▀
;; (code-char 9620) => #\UPPER_ONE_EIGHTH_BLOCK     <=> #\▔
;; (code-char 9601) => #\LOWER_ONE_EIGHTH_BLOCK     <=> #\▁
;; (code-char 9602) => #\LOWER_ONE_QUARTER_BLOCK    <=> #\▂
;; (code-char 9603) => #\LOWER_THREE_EIGHTHS_BLOCK  <=> #\▃
;; (code-char 9604) => #\LOWER_HALF_BLOCK           <=> #\▄
;; (code-char 9605) => #\LOWER_FIVE_EIGHTHS_BLOCK   <=> #\▅
;; (code-char 9606) => #\LOWER_THREE_QUARTERS_BLOCK <=> #\▆
;; (code-char 9607) => #\LOWER_SEVEN_EIGHTHS_BLOCK  <=> #\▇
;; (code-char 9608) => #\FULL_BLOCK                 <=> #\█
;; (code-char 9135) => #\HORIZONTAL_LINE_EXTENSION  <=> #\⎯

(defvar *ticks*
  (vector (code-char 9601) (code-char 9602) (code-char 9603)
          (code-char 9604) (code-char 9605) (code-char 9606)
          (code-char 9607) (code-char 9608))
  "
A simple-vector of characters for representation of sparklines.
Default is #(#\▁ #\▂ #\▃ #\▄ #\▅ #\▆ #\▇ #\█).

Examples:

  (defvar ternary '(-1 0 1 -1 1 0 -1 1 -1))

  (spark ternary)              => \"▁▄█▁█▄▁█▁\"

  (let ((*ticks* #(#\_ #\- #\¯)))
    (spark ternary))           => \"_-¯_¯-_¯_\"

  (let ((*ticks* #(#\▄ #\⎯ #\▀)))
    (spark ternary))           => \"▄⎯▀▄▀⎯▄▀▄\"
")


(defun spark (numbers &key min max key)
  (check-type numbers list)
  (check-type min     (or null real))
  (check-type max     (or null real))
  (check-type key     (or symbol function))
  (when key (setf numbers (mapcar key numbers)))

  ;; Empty data case:
  (when (null numbers)
    (RETURN-FROM spark ""))

  ;; Ensure min is the minimum number.
  (if (null min)
      (setf min (reduce #'min numbers))
      (setf numbers (mapcar (lambda (n) (max n min)) numbers)))

  ;; Ensure max is the maximum number.
  (if (null max)
      (setf max (reduce #'max numbers))
      (setf numbers (mapcar (lambda (n) (min n max)) numbers)))

  (when (< max min)
    (error "max ~S < min ~S." max min))

  (let ((unit (/ (- max min) (1- (length *ticks*)))))
    (when (zerop unit) (setf unit 1))
    (with-output-to-string (s)
      (loop :for n :in numbers
            :for nth := (floor (- n min) unit)
            :do (princ (svref *ticks* nth) s)))))


(setf (documentation 'spark 'function) "
Generates a sparkline string for a list of real numbers.

Usage: SPARK <numbers> &key <min> <max> <key>

  * <numbers> ::= <list> of <real-number>
  * <min>     ::= { <null> | <real-number> }, default is NIL
  * <max>     ::= { <null> | <real-number> }, default is NIL
  * <key>     ::= <function>

  * <numbers> ~ data.
  * <min>    ~ lower bound of output.
               NIL means the minimum value of the data.
  * <max>    ~ upper bound of output.
               NIL means the maximum value of the data.
  * <key>    ~ function for preparing data.

Examples:

  (spark '(1 0 1 0))     => \"█▁█▁\"
  (spark '(1 0 1 0 0.5)) => \"█▁█▁▄\"
  (spark '(1 0 1 0 -1))  => \"█▄█▄▁\"

  (spark '(0 30 55 80 33 150))                 => \"▁▂▃▅▂█\"
  (spark '(0 30 55 80 33 150) :min -100)       => \"▃▄▅▆▄█\"
  (spark '(0 30 55 80 33 150) :max 50)         => \"▁▅██▅█\"
  (spark '(0 30 55 80 33 150) :min 30 :max 80) => \"▁▁▄█▁█\"

  (spark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4))))
  => \"▄▆█▆▄▂▁▂▄\"
  (spark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (cos (* x pi 1/4))))
  => \"█▆▄▂▁▂▄▆█\"

 For more examples, see cl-spark/spark-test.lisp
")


;;--------------------------------------------------------------------
;; Vspark
;;--------------------------------------------------------------------
;; #(#\▏ #\▎ #\▍ #\▌ #\▋ #\▊ #\▉ #\█)

;; (code-char 9615) => #\LEFT_ONE_EIGHTH_BLOCK     <=> #\▏
;; (code-char 9614) => #\LEFT_ONE_QUARTER_BLOCK    <=> #\▎
;; (code-char 9613) => #\LEFT_THREE_EIGHTHS_BLOCK  <=> #\▍
;; (code-char 9612) => #\LEFT_HALF_BLOCK           <=> #\▌
;; (code-char 9611) => #\LEFT_FIVE_EIGHTHS_BLOCK   <=> #\▋
;; (code-char 9610) => #\LEFT_THREE_QUARTERS_BLOCK <=> #\▊
;; (code-char 9609) => #\LEFT_SEVEN_EIGHTHS_BLOCK  <=> #\▉
;; (code-char 9608) => #\FULL_BLOCK                <=> #\█
;; (code-char 9616) => #\RIGHT_HALF_BLOCK          <=> #\▐
;; (code-char 9621) => #\RIGHT_ONE_EIGHTH_BLOCK    <=> #\▕

(defvar *vticks*
  (vector (code-char 9615) (code-char 9614) (code-char 9613)
          (code-char 9612) (code-char 9611) (code-char 9610)
          (code-char 9609) (code-char 9608))
  "
A simple-vector of characters for representation of vartical
sparklines. Default is #(#\▏ #\▎ #\▍ #\▌ #\▋ #\▊ #\▉ #\█).

Examples:

  ;; Japan GDP growth rate, annal
  ;; see. http://data.worldbank.org/indicator/NY.GDP.MKTP.KD.ZG
  (defparameter growth-rate
   '((2007 . 2.192186) (2008 . -1.041636) (2009 . -5.5269766)
     (2010 . 4.652112) (2011 . -0.57031655) (2012 . 1.945)))

  (vspark growth-rate :key #'cdr :labels (mapcar #'car growth-rate))
  =>
  \"
       -5.5269766        -0.4374323         4.652112
       ˫---------------------+---------------------˧
  2007 ██████████████████████████████████▏
  2008 ███████████████████▊
  2009 ▏
  2010 ████████████████████████████████████████████
  2011 █████████████████████▉
  2012 █████████████████████████████████▏
  \"

  (let ((*vticks* #(#\- #\0 #\+)))
    (vspark growth-rate :key (lambda (y-r) (float-sign (cdr y-r)))
                        :labels (mapcar #'car growth-rate)
                        :size 1))
  =>
  \"
  2007 +
  2008 -
  2009 -
  2010 +
  2011 -
  2012 +
  \"
")

(defun vspark (numbers &key min max key (size 50) labels title (scale? t) (newline? t))
  (check-type numbers  list)
  (check-type min      (or null real))
  (check-type max      (or null real))
  (check-type key      (or symbol function))
  (check-type size     (integer 1 *))
  (check-type labels   list)

  (when key (setf numbers (mapcar key numbers)))

  ;; Empty data case:
  (when (null numbers)
    (RETURN-FROM vspark ""))

  ;; Ensure min is the minimum number.
  (if (null min)
      (setf min (reduce #'min numbers))
      (setf numbers (mapcar (lambda (n) (max n min)) numbers)))

  ;; Ensure max is the maximum number.
  (if (null max)
      (setf max (reduce #'max numbers))
      (setf numbers (mapcar (lambda (n) (min n max)) numbers)))

  ;; Check max ~ min.
  (cond ((< max min) (error "max ~S < min ~S." max min))
        ((= max min) (incf max))        ; ensure all bars are in min.
        (t nil))

  (let ((max-lengeth-label nil))
    (when labels
      ;; Ensure num labels equals to num numbers.
      (let ((diff (- (length numbers) (length labels))))
        (cond ((plusp diff)
               ;; Add padding lacking labels not to miss data.
               (setf labels (append labels (loop :repeat diff :collect ""))))
              ((minusp diff)
               ;; Remove superfluous labels to remove redundant spaces.
               (setf labels (butlast labels (abs diff))))
              (t nil)))
      ;; Find max-lengeth-label.
      (setf max-lengeth-label
            (reduce #'max labels
                    :key (lambda (label)
                           (if (stringp label)
                               (length label)
                               (length (format nil "~A" label))))))
      ;; Canonicalize labels.
      (let* ((control-string (format nil "~~~D,,@A " max-lengeth-label)))
        (setf labels
              (mapcar (lambda (label) (format nil control-string label))
                      labels)))
      ;; Reduce size for max-lengeth-label.
      ;;  * 1 is space between label and bar
      ;;  * ensure minimum size 1
      (setf size (max 1 (- size 1 max-lengeth-label))))

    (let* ((num-content-ticks (1- (length *vticks*)))
           (unit (/ (- max min) (* size num-content-ticks)))
           (result '()))
      (when (zerop unit) (setf unit 1))

      (loop :for n :in numbers
            :for i :from 0
            :do (when labels (push (nth i labels) result))
                (push (generate-bar n unit min max num-content-ticks)
                      result)
            :finally (setf result (nreverse result)))

      (when scale?
        (awhen (generate-scale min max size max-lengeth-label)
          (push it result)))

      (when title
        (awhen (generate-title title size max-lengeth-label)
          (push it result)))

      (if newline?
          (apply #'string-concat (push #.(format nil "~%") result))
          (string-right-trim '(#\Newline)
                             (apply #'string-concat result))))))


(setf (documentation 'vspark 'function) "
Generates a vartical sparkline string for a list of real numbers.

Usage: VSPARK <numbers> &key <min> <max> <key> <size>
                             <labels> <title> <scale?> <newline?>

  * <numbers>  ::= <list> of <real-number>
  * <min>      ::= { <null> | <real-number> }, default is NIL
  * <max>      ::= { <null> | <real-number> }, default is NIL
  * <key>      ::= <function>
  * <size>     ::= <integer 1 *>, default is 50
  * <labels>   ::= <list>
  * <title>    ::= <object>, default is NIL
  * <scale?>   ::= <generalized-boolean>, default is T
  * <newline?> ::= <generalized-boolean>, default is T

  * <numbers>  ~ data.
  * <min>      ~ lower bound of output.
                 NIL means the minimum value of the data.
  * <max>      ~ upper bound of output.
                 NIL means the maximum value of the data.
  * <key>      ~ function for preparing data.
  * <size>     ~ maximum number of output columns (contains label).
  * <labels>   ~ labels for data.
  * <title>    ~ If title is too big for size, it is not printed.
  * <scale?>   ~ If T, output graph with scale for easy to see.
                 If string length of min and max is too big for size,
                 the scale is not printed.
  * <newline?> ~ If T, output graph with newlines for easy to see.


Examples:

  ;; Life expectancy by WHO region, 2011, bothsexes
  ;; see. http://apps.who.int/gho/data/view.main.690
  (defvar life-expectancies '((\"Africa\" 56)
                              (\"Americans\" 76)
                              (\"South-East Asia\" 67)
                              (\"Europe\" 76)
                              (\"Eastern Mediterranean\" 68)
                              (\"Western Pacific\" 76)
                              (\"Global\" 70)))

  (vspark life-expectancies :key #'second :scale? nil :newline? nil)
  =>
  \"▏
  ██████████████████████████████████████████████████
  ███████████████████████████▌
  ██████████████████████████████████████████████████
  ██████████████████████████████▏
  ██████████████████████████████████████████████████
  ███████████████████████████████████▏\"

  (vspark life-expectancies :min 50 :max 80
                            :key    #'second
                            :labels (mapcar #'first life-expectancies)
                            :title \"Life Expectancy\")
  =>
  \"
                   Life Expectancy                  
                        50           65           80
                        ˫------------+-------------˧
                 Africa █████▋
              Americans ████████████████████████▎
        South-East Asia ███████████████▉
                 Europe ████████████████████████▎
  Eastern Mediterranean ████████████████▊
        Western Pacific ████████████████████████▎
                 Global ██████████████████▋
  \"

  (vspark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4)))
                               :size 20)
  \"
  -1.0     0.0     1.0
  ˫--------+---------˧
  ██████████▏
  █████████████████▏
  ████████████████████
  █████████████████▏
  ██████████▏
  ██▉
  ▏
  ██▉
  █████████▉
  \"

  (vspark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4)))
                               :size 10)
  =>
  \"
  -1.0   1.0
  ˫--------˧
  █████▏
  ████████▏
  ██████████
  ████████▏
  █████▏
  █▏
  ▏
  █▏
  ████▏
  \"

  (vspark '(0 1 2 3 4 5 6 7 8) :key (lambda (x) (sin (* x pi 1/4)))
                               :size 1)
  =>
  \"
  ▌
  ▊
  █
  ▊
  ▌
  ▎
  ▏
  ▎
  ▌
  \"

  For more examples, see cl-spark/spark-test.lisp
")

(defun generate-bar (number unit min max num-content-ticks)
  (multiple-value-bind
        (units frac) (floor (- number min) (* unit num-content-ticks))
    (with-output-to-string (s)
      (let ((most-tick (svref *vticks* num-content-ticks)))
        (dotimes (i units) (princ most-tick s))
        (unless (= number max)
          ;; max number need not frac.
          ;; if number = max, then always frac = 0.
          (princ (svref *vticks* (floor frac unit))
                 s))
        (terpri s)))))

(defun generate-title (title size max-lengeth-label)
  (let* ((title-string (princ-to-string title))
         (mid (floor (- (if max-lengeth-label
                            (+ 1 size max-lengeth-label)
                            size)
                        (length title-string)) 2)))
    (when (plusp mid)
      (format nil "~A~%"
              (replace (make-string (if max-lengeth-label
                                        (+ 1 size max-lengeth-label)
                                        size)
                                    :initial-element #\Space)
                       title-string :start1 mid)))))

;; (code-char 743) => #\MODIFIER_LETTER_MID_TONE_BAR              <=> #\˧
;; (code-char 746) => #\MODIFIER_LETTER_YANG_DEPARTING_TONE_MARK  <=> #\˫
(defun generate-scale (min max size max-lengeth-label)
  (flet ((to-string (n)
           (princ-to-string (if (integerp n) n (float n 0.0)))))
   (let* ((min-string (to-string min))
          (max-string (to-string max))
          (num-pad    (- size (length min-string) (length max-string))))
     (when (plusp num-pad)
       (let* ((mid        (/ (+ max min) 2))
              (mid-string (to-string mid))
              (num-indent (aif max-lengeth-label (1+ it) 0))
              (num-pad2   (- num-pad (length mid-string))))
         (if (and (plusp num-pad2)
                  (/= min mid)
                  (/= mid max))
             ;; A. mid exist case:
             (with-output-to-string (s)
               (dotimes (_ num-indent) (princ #\Space s))
               (princ min-string s)
               (dotimes (_ (ceiling num-pad2 2)) (princ #\Space s))
               (princ mid-string s)
               (dotimes (_ (floor num-pad2 2)) (princ #\Space s))
               (princ max-string s)
               (princ #\Newline s)
               (dotimes (_ num-indent) (princ #\Space s))
               (princ #.(code-char 747) s)
               (dotimes (_ (ceiling (- size 3) 2)) (princ #\- s))
               (princ #\+ s)                                                      
               (dotimes (_ (floor (- size 3) 2)) (princ #\- s))
               (princ #.(code-char 743) s)                           
               (princ #\Newline s))

             ;; B. no mid exist case:
             (with-output-to-string (s)
               (dotimes (_ num-indent) (princ #\Space s))
               (princ min-string s)
               (dotimes (_ num-pad) (princ #\Space s))
               (princ max-string s)
               (princ #\Newline s)
               (dotimes (_ num-indent) (princ #\Space s))
               (princ #.(code-char 747) s)
               (dotimes (_ (- size 2)) (princ #\- s))
               (princ #.(code-char 743) s)                           
               (princ #\Newline s))))))))


;;====================================================================
